apps/trivium/src/static_deque/static_bit_deque.rs

```diff
@@ -0,0 +1,136 @@
+//! This module implements the StaticBitDeque struct: a deque of bytes. The idea
+//! is that this is a wrapper around StaticDeque, but StaticBitDeque has an additional
+//! functionnality: it can construct the "intermediate" bytes, made of parts of other bytes.
+//! This is pretending to store bits, and allows accessing bits in chunks of 8 consecutive.
+
+use crate::static_deque::StaticDeque;
+
+use tfhe::FheUint8;
+
+/// Internal trait specifying which operations are needed by StaticBitDeque
+pub trait StaticBitDequeInput<OpOutput>:
+    Clone
+    + std::ops::Shr<u8, Output = OpOutput>
+    + std::ops::Shl<u8, Output = OpOutput>
+    + std::ops::BitOr<Output = OpOutput>
+{
+}
+impl StaticBitDequeInput<u8> for u8 {}
+impl StaticBitDequeInput<u8> for &u8 {}
+impl StaticBitDequeInput<FheUint8> for FheUint8 {}
+impl StaticBitDequeInput<FheUint8> for &FheUint8 {}
+
+/// Here T must represent a type covering a byte, like u8 or FheUint8.
+#[derive(Clone)]
+pub struct StaticBitDeque<const N: usize, T> {
+    deque: StaticDeque<N, T>,
+}
+
+impl<const N: usize, T> StaticBitDeque<N, T>
+where
+    T: StaticBitDequeInput<T>,
+    for<'a> &'a T: StaticBitDequeInput<T>,
+{
+    /// Constructor always uses a fully initialized array, the first element of
+    /// which is oldest, the last is newest
+    pub fn new(_arr: [T; N]) -> Self {
+        Self {
+            deque: StaticDeque::<N, T>::new(_arr),
+        }
+    }
```

Comment:
put an assert on the size of T here ? as it is required for it to be a byte-like object with an explicit error message

Reply:
But if T can be a FheUint8, it can have any size, doesn't it ?

Also the byte-like object is a soft requirement: in practice you only need a type that implements some bitshift operations

Reply:
Indeed was a bit too optimistic on that assert haha